src/storage/slab/item.c

```diff
@@ -372,3 +374,76 @@ item_flush(void)
     flush_at = time_now();
     log_info("all keys flushed at %"PRIu32, flush_at);
 }
+
+void
+item_count(size_t *nkey, size_t *ksize, size_t *vsize, struct bstring *prefix)
+{
+    uint32_t nbucket = HASHSIZE(hash_table->hash_power);
+
+    log_info("start scanning all %"PRIu32" keys", hash_table->nhash_item);
+
+    *nkey = 0;
+    *ksize = 0;
+    *vsize = 0;
+    for (uint32_t i = 0; i < nbucket; i++) {
+        struct item_slh *entry = &hash_table->table[i];
+        struct item *it;
+
+        SLIST_FOREACH(it, entry, i_sle) {
+            if (it->klen >= prefix->len &&
+                    cc_bcmp(prefix->data, item_key(it), prefix->len) == 0) {
+                *nkey += 1;
+                *ksize += it->klen;
+                *vsize += it->vlen;
+            }
+        }
+
+        if (i % 1000000 == 0) {
+            log_info("... %"PRIu32" out of %"PRIu32" buckets scanned ...", i,
+                    nbucket);
+        }
+    }
+
+    log_info("finish scanning all keys");
+}
+
+bool
+item_dump(void)
+{
+    int fd;
+    uint32_t nbucket = HASHSIZE(hash_table->hash_power);
+
+    log_info("start scanning all %"PRIu32" keys", hash_table->nhash_item);
+
+    fd = open("key.dump", O_WRONLY | O_TRUNC | O_CREAT, 0644);
+    if (fd < 0) {
+        log_stderr("Could not create key dump - cannot open file");
+        return false;
+    }
+
+    for (uint32_t i = 0; i < nbucket; i++) {
+        struct item_slh *entry = &hash_table->table[i];
+        struct item *it;
+
+        SLIST_FOREACH(it, entry, i_sle) {
+            if (write(fd, item_key(it), it->klen) < it->klen) {
```

Comment:
would we also want to check for expiry?

Reply:
I guess there're reasons to do either/both.

The only case I can think of where someone wants expired keys to be dumped, is to investigate how much space in cache is occupied by already expired keys. But that is a legitimate question that we have been asked.

I think we can keep this change simple, while clarifying what the current logic does. In the future, if someone requests dumping only live objects, we can make it an option, e.g. "dump live" "dump expired" "dump all"